File: src/geometry_right_triangle_trig_easy.py
"""
geometry - right_triangle_trig (easy)
Generated: 2026-02-22T04:18:52.199856
"""

from problem_utils import *
from svg_utils import Diagram

# Special angles with exact trig values
SPECIAL_ANGLES = {
    30: {'sin': Rational(1,2), 'cos': sqrt(3)/2, 'tan': 1/sqrt(3)},
    45: {'sin': sqrt(2)/2, 'cos': sqrt(2)/2, 'tan': Integer(1)},
    60: {'sin': sqrt(3)/2, 'cos': Rational(1,2), 'tan': sqrt(3)},
}

def generate():
    problem_type = randint(1, 4)

    if problem_type == 1:
        # Find a side using sin/cos/tan with a special angle
        angle = choice([30, 45, 60])
        trig_name = choice(['sin', 'cos', 'tan'])
        scale = choice([2, 3, 4, 5, 6, 8, 10, 12, 15, 20])
        ratios = SPECIAL_ANGLES[angle]
        ratio_val = ratios[trig_name]

        if trig_name == 'sin':
            hyp = scale
            opp = simplify(ratio_val * hyp)
            question_text = f"In a right triangle, if the hypotenuse is ${hyp}$ and one angle is ${angle}°$, find the side opposite to the ${angle}°$ angle."
            ans = opp
            sol = steps(
                f"Use $\\sin({angle}°) = \\frac{{\\text{{opposite}}}}{{\\text{{hypotenuse}}}}$",
                f"$\\sin({angle}°) = {latex(ratio_val)}$",
                f"$\\text{{opposite}} = {hyp} \\cdot {latex(ratio_val)} = {latex(ans)}$"
            )
        elif trig_name == 'cos':
            hyp = scale
            adj = simplify(ratio_val * hyp)
            question_text = f"In a right triangle, if the hypotenuse is ${hyp}$ and one angle is ${angle}°$, find the side adjacent to the ${angle}°$ angle."
            ans = adj
            sol = steps(
                f"Use $\\cos({angle}°) = \\frac{{\\text{{adjacent}}}}{{\\text{{hypotenuse}}}}$",
                f"$\\cos({angle}°) = {latex(ratio_val)}$",
                f"$\\text{{adjacent}} = {hyp} \\cdot {latex(ratio_val)} = {latex(ans)}$"
            )
        else:  # tan
            adj = scale
            opp = simplify(ratio_val * adj)
            question_text = f"In a right triangle, if the side adjacent to a ${angle}°$ angle is ${adj}$, find the side opposite to the ${angle}°$ angle."
            ans = opp
            sol = steps(
                f"Use $\\tan({angle}°) = \\frac{{\\text{{opposite}}}}{{\\text{{adjacent}}}}$",
                f"$\\tan({angle}°) = {latex(ratio_val)}$",
                f"$\\text{{opposite}} = {adj} \\cdot {latex(ratio_val)} = {latex(ans)}$"
            )

        return problem(
            question=question_text,
            answer=simplify(ans),
            difficulty=(1000, 1200),
            topic="geometry/right_triangle_trig",
            solution=sol
        )

    elif problem_type == 2:
        # Find an angle given a ratio that matches a special angle
        angle = choice([30, 45, 60])
        trig_name = choice(['sin', 'cos', 'tan'])
        ratio_val = SPECIAL_ANGLES[angle][trig_name]
        scale = randint(1, 6)

        if trig_name == 'sin':
            opp = simplify(ratio_val * scale * 2) if angle in [30,60] else simplify(ratio_val * scale)
            hyp = simplify(scale * 2) if angle in [30,60] else simplify(sqrt(2) * scale)
            question_text = f"In a right triangle, the side opposite an angle is ${latex(simplify(ratio_val * scale * 2 if angle in [30,60] else ratio_val*scale))}$ and the hypotenuse is ${latex(simplify(2*scale if angle in [30,60] else sqrt(2)*scale))}$. Find the angle in degrees."
            sol = steps(
                f"$\\sin(\\theta) = \\frac{{\\text{{opposite}}}}{{\\text{{hypotenuse}}}} = {latex(ratio_val)}$",
                f"$\\theta = \\sin^{{-1}}({latex(ratio_val)}) = {angle}°$"
            )
        elif trig_name == 'cos':
            adj = simplify(ratio_val * scale * 2) if angle in [30,60] else simplify(ratio_val * scale)
            hyp = simplify(scale * 2) if angle in [30,60] else simplify(sqrt(2) * scale)
            question_text = f"In a right triangle, the side adjacent to an angle is ${latex(simplify(ratio_val * scale * 2 if angle in [30,60] else ratio_val*scale))}$ and the hypotenuse is ${latex(simplify(2*scale if angle in [30,60] else sqrt(2)*scale))}$. Find the angle in degrees."
            sol = steps(
                f"$\\cos(\\theta) = \\frac{{\\text{{adjacent}}}}{{\\text{{hypotenuse}}}} = {latex(ratio_val)}$",
                f"$\\theta = \\cos^{{-1}}({latex(ratio_val)}) = {angle}°$"
            )
        else:
            opp_scale = scale
            adj_scale = simplify(opp_scale / ratio_val)
            question_text = f"In a right triangle, the side opposite an angle is ${opp_scale}$ and the side adjacent is ${latex(simplify(adj_scale))}$. Find the angle in degrees."
            sol = steps(
                f"$\\tan(\\theta) = \\frac{{\\text{{opposite}}}}{{\\text{{adjacent}}}} = \\frac{{{opp_scale}}}{{{latex(simplify(adj_scale))}}} = {latex(ratio_val)}$",
                f"$\\theta = \\tan^{{-1}}({latex(ratio_val)}) = {angle}°$"
            )

        return problem(
            question=question_text,
            answer=angle,
            difficulty=(1100, 1250),
            topic="geometry/right_triangle_trig",
            solution=sol
        )

    elif problem_type == 3:
        # SOH-CAH-TOA identification: write trig ratio as fraction
        angle_label = choice(['A', 'B', 'C', 'D', 'E'])
        # Pick non-equal coprime sides
        side1 = randint(2, 15)
        side2 = randint(2, 15)
        while side1 == side2:
            side2 = randint(2, 15)
        ratio_type = choice(['sin', 'cos', 'tan'])
        ans = Rational(side1, side2)

        if ratio_type == 'sin':
            q = f"In a right triangle, the side opposite angle ${angle_label}$ is ${side1}$ and the hypotenuse is ${side2}$. Write $\\sin({angle_label})$ as a fraction."
            sol = steps(
                f"$\\sin({angle_label}) = \\frac{{\\text{{opposite}}}}{{\\text{{hypotenuse}}}} = \\frac{{{side1}}}{{{side2}}}$"
            )
        elif ratio_type == 'cos':
            q = f"In a right triangle, the side adjacent to angle ${angle_label}$ is ${side1}$ and the hypotenuse is ${side2}$. Write $\\cos({angle_label})$ as a fraction."
            sol = steps(
                f"$\\cos({angle_label}) = \\frac{{\\text{{adjacent}}}}{{\\text{{hypotenuse}}}} = \\frac{{{side1}}}{{{side2}}}$"
            )
        else:
            q = f"In a right triangle, the side opposite angle ${angle_label}$ is ${side1}$ and the side adjacent is ${side2}$. Write $\\tan({angle_label})$ as a fraction."
            sol = steps(
                f"$\\tan({angle_label}) = \\frac{{\\text{{opposite}}}}{{\\text{{adjacent}}}} = \\frac{{{side1}}}{{{side2}}}$"
            )

        return problem(
            question=q,
            answer=ans,
            difficulty=(1000, 1150),
            topic="geometry/right_triangle_trig",
            solution=sol
        )

    else:
        # 30-60-90 and 45-45-90 triangles, given hypotenuse find legs
        triangle_type = choice(['30-60-90', '45-45-90'])
        hypotenuse = choice([2, 4, 6, 8, 10, 12, 14, 16, 20])

        if triangle_type == '30-60-90':
            short_leg = simplify(hypotenuse * Rational(1,2))
            long_leg = simplify(hypotenuse * sqrt(3) / 2)
            which = choice(['short', 'long'])
            if which == 'short':
                ans = short_leg
                q = f"In a $30°$-$60°$-$90°$ triangle, the hypotenuse is ${hypotenuse}$. Find the length of the shorter leg (opposite $30°$)."
                sol = steps(
                    f"In a $30°$-$60°$-$90°$ triangle, the sides are in ratio $1:\\sqrt{{3}}:2$",
                    f"Shorter leg $= \\frac{{\\text{{hypotenuse}}}}{{2}} = \\frac{{{hypotenuse}}}{{2}} = {latex(short_leg)}$"
                )
            else:
                ans = long_leg
                q = f"In a $30°$-$60°$-$90°$ triangle, the hypotenuse is ${hypotenuse}$. Find the length of the longer leg (opposite $60°$)."
                sol = steps(
                    f"In a $30°$-$60°$-$90°$ triangle, sides are in ratio $1:\\sqrt{{3}}:2$",
                    f"Longer leg $= \\frac{{\\sqrt{{3}}}}{{2}} \\cdot {hypotenuse} = {latex(long_leg)}$"
                )
        else:
            leg = simplify(hypotenuse * sqrt(2) / 2)
            ans = leg
            q = f"In a $45°$-$45°$-$90°$ triangle, the hypotenuse is ${hypotenuse}$. Find the length of one leg."
            sol = steps(
                f"In a $45°$-$45°$-$90°$ triangle, the sides are in ratio $1:1:\\sqrt{{2}}$",
                f"Leg $= \\frac{{\\text{{hypotenuse}}}}{{\\sqrt{{2}}}} = \\frac{{{hypotenuse}}}{{\\sqrt{{2}}}} = {latex(leg)}$"
            )

        return problem(
            question=q,
            answer=ans,
            difficulty=(1200, 1300),
            topic="geometry/right_triangle_trig",
            solution=sol
        )

emit(generate())

File: src/geometry_right_triangle_trig_easy.jl
# geometry - right_triangle_trig (easy)
# Generated: 2026-03-08T20:12:33.399910

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("geometry/right_triangle_trig")
    
    # Choose problem type
    prob_type = choice([:solve_side, :solve_angle, :find_trig_ratio])
    
    if prob_type == :solve_side
        # Given angle and one side, find another side
        angle_deg = choice([30, 45, 60])
        side_type = choice([:opp_given, :adj_given, :hyp_given])
        
        if angle_deg == 30
            # Use 30-60-90 triangle ratios
            if side_type == :opp_given
                opp = randint(3, 12)
                adj = round(Int, opp * sqrt(3))
                hyp = 2 * opp
                given_side = opp
                given_label = "opposite"
                find_side = "hypotenuse"
                answer = hyp
            elseif side_type == :adj_given
                adj = randint(3, 12)
                opp = round(Int, adj / sqrt(3))
                hyp = round(Int, 2 * adj / sqrt(3))
                given_side = adj
                given_label = "adjacent"
                find_side = "opposite"
                answer = opp
            else
                hyp = randint(6, 24)
                opp = hyp ÷ 2
                adj = round(Int, opp * sqrt(3))
                given_side = hyp
                given_label = "hypotenuse"
                find_side = "adjacent"
                answer = adj
            end
        elseif angle_deg == 45
            # Use 45-45-90 triangle ratios
            if side_type == :opp_given
                opp = randint(4, 15)
                adj = opp
                hyp = round(Int, opp * sqrt(2))
                given_side = opp
                given_label = "opposite"
                find_side = "hypotenuse"
                answer = hyp
            elseif side_type == :adj_given
                adj = randint(4, 15)
                opp = adj
                hyp = round(Int, adj * sqrt(2))
                given_side = adj
                given_label = "adjacent"
                find_side = "opposite"
                answer = opp
            else
                hyp = randint(8, 20)
                opp = round(Int, hyp / sqrt(2))
                adj = opp
                given_side = hyp
                given_label = "hypotenuse"
                find_side = "adjacent"
                answer = adj
            end
        else  # 60 degrees
            # Use 30-60-90 triangle ratios (60° is the larger acute angle)
            if side_type == :opp_given
                opp = randint(4, 15)
                adj = round(Int, opp / sqrt(3))
                hyp = round(Int, 2 * opp / sqrt(3))
                given_side = opp
                given_label = "opposite"
                find_side = "adjacent"
                answer = adj
            elseif side_type == :adj_given
                adj = randint(3, 12)
                opp = round(Int, adj * sqrt(3))
                hyp = 2 * adj
                given_side = adj
                given_label = "adjacent"
                find_side = "hypotenuse"
                answer = hyp
            else
                hyp = randint(6, 24)
                adj = hyp ÷ 2
                opp = round(Int, adj * sqrt(3))
                given_side = hyp
                given_label = "hypotenuse"
                find_side = "opposite"
                answer = opp
            end
        end
        
        # Create diagram
        d = DiagramObj(width=300, height=300)
        # Draw right triangle
        polygon!(d, [(50, 250), (250, 250), (50, 100)], fill="lightblue", stroke="black", stroke_width=2)
        # Mark right angle
        right_angle!(d, (50, 250), (250, 250), (50, 100), size=15)
        # Mark the given angle
        if angle_deg != 90
            angle_arc!(d, 50, 100, 270, 270+angle_deg, 30, label="$(angle_deg)°")
        end
        # Label sides
        opp_label = given_label == "opposite" ? string(given_side) : find_side == "opposite" ? "?" : ""
        adj_label = given_label == "adjacent" ? string(given_side) : find_side == "adjacent" ? "?" : ""
        hyp_label = given_label == "hypotenuse" ? string(given_side) : find_side == "hypotenuse" ? "?" : ""
        segment_label!(d, 50, 175, opp_label)
        segment_label!(d, 150, 260, adj_label)
        segment_label!(d, 150, 170, hyp_label)
        
        question = "In a right triangle, one angle measures $(angle_deg)°. The side $(given_label) to this angle has length $(given_side). Find the length of the $(find_side)."
        
        if angle_deg == 30 || angle_deg == 60
            trig_ratio = angle_deg == 30 ? (side_type == :opp_given ? "sin" : side_type == :adj_given ? "cos" : "sin") : (side_type == :opp_given ? "sin" : side_type == :adj_given ? "tan" : "cos")
        else
            trig_ratio = side_type == :opp_given ? "sin" : side_type == :adj_given ? "cos" : "sin"
        end
        
        solution_text = steps(
            "Given: angle = $(angle_deg)°, $(given_label) = $(given_side)",
            "Use the appropriate trigonometric ratio for a $(angle_deg)° angle in a right triangle",
            "Apply special triangle ratios or trigonometric functions",
            sol("Answer", answer)
        )
        
        problem(
            question=question,
            answer=answer,
            difficulty=(800, 1000),
            solution=solution_text,
            image=render(d),
            time=90
        )
        
    elseif prob_type == :solve_angle
        # Given two sides, find an angle
        opp = randint(3, 12)
        adj = randint(3, 12)
        hyp = round(Int, sqrt(opp^2 + adj^2))
        
        # Create diagram
        d = DiagramObj(width=300, height=300)
        polygon!(d, [(50, 250), (250, 250), (50, 100)], fill="lightblue", stroke="black", stroke_width=2)
        right_angle!(d, (50, 250), (250, 250), (50, 100), size=15)
        segment_label!(d, 50, 175, "$(opp)")
        segment_label!(d, 150, 260, "$(adj)")
        angle_arc!(d, 50, 100, 270, 315, 30, label="θ")
        
        # Calculate angle using arctan
        angle_rad = atan(opp / adj)
        angle_deg = round(angle_rad * 180 / π, digits=1)
        
        question = "In a right triangle, the opposite side has length $(opp) and the adjacent side has length $(adj). Find the angle θ (in degrees, rounded to one decimal place)."
        
        solution_text = steps(
            "Given: opposite = $(opp), adjacent = $(adj)",
            "Use tan(θ) = opposite/adjacent = $(opp)/$(adj)",
            "θ = arctan($(opp)/$(adj))",
            sol("Answer", "$(angle_deg)°")
        )
        
        problem(
            question=question,
            answer=angle_deg,
            difficulty=(900, 1100),
            solution=solution_text,
            image=render(d),
            calculator="scientific",
            time=120
        )
        
    else  # find_trig_ratio
        # Given a right triangle with sides, find a trigonometric ratio
        a = randint(3, 12)
        b = randint(3, 12)
        c = round(Int, sqrt(a^2 + b^2))
        
        # Create diagram
        d = DiagramObj(width=300, height=300)
        polygon!(d, [(50, 250), (250, 250), (50, 100)], fill="lightblue", stroke="black", stroke_width=2)
        right_angle!(d, (50, 250), (250, 250), (50, 100), size=15)
        segment_label!(d, 50, 175, "$(a)")
        segment_label!(d, 150, 260, "$(b)")
        segment_label!(d, 150, 170, "$(c)")
        angle_arc!(d, 50, 100, 270, 315, 30, label="θ")
        
        trig_func = choice(["sin", "cos", "tan"])
        
        if trig_func == "sin"
            answer = a // c
            ratio_text = "opposite/hypotenuse = $(a)/$(c)"
        elseif trig_func == "cos"
            answer = b // c
            ratio_text = "adjacent/hypotenuse = $(b)/$(c)"
        else
            answer = a // b
            ratio_text = "opposite/adjacent = $(a)/$(b)"
        end
        
        question = "In the right triangle shown, the sides have lengths $(a), $(b), and $(c). Find $(trig_func)(θ) as a fraction in lowest terms."
        
        solution_text = steps(
            "Identify the sides relative to angle θ: opposite = $(a), adjacent = $(b), hypotenuse = $(c)",
            "$(trig_func)(θ) = $(ratio_text)",
            "Simplify the fraction if possible",
            sol("Answer", answer)
        )
        
        problem(
            question=question,
            answer=answer,
            difficulty=(700, 900),
            solution=solution_text,
            image=render(d),
            time=90
        )
    end
end